# Underwriter vs Lending Club

`int_rate` is set by Lending Club and reflects its risk assessment. Including it measures
performance when that assessment is available, while excluding it measures performance from
borrower data alone.

This notebook compares three feature sets: Lending Club's verdict (`int_rate` and `grade`),
the borrower data available at application, and their union.

What it shows:

- For LightGBM, the union performs better than either feature set alone across all three metrics, with confidence intervals supporting these improvements. Borrower data alone has higher PR AUC than the Lending Club verdict, while their differences in ROC AUC and Brier score remain uncertain.
- LightGBM has higher ROC AUC and PR AUC than logistic regression on the borrower and union
  sets, while logistic regression scores higher on these metrics for the verdict set, which
  consists mainly of one monotonic feature.
- The borrower features explain about 40% of the variation in `int_rate`, so pricing also relies
  on information absent from this dataset.
- FICO and DTI continue to separate bad rates within a single grade, providing information that
  the union model can use.
- FICO has the highest permutation importance for both models. The engineered ratios contribute
  more to logistic regression, while the derogatory-event counts contribute little to either.

In [1]:
from collections.abc import Callable

import numpy as np
import pandas as pd
from numpy.typing import NDArray

from IPython.display import display
from lightgbm import LGBMRegressor
from sklearn.inspection import permutation_importance
from sklearn.metrics import r2_score
from sklearn.pipeline import Pipeline

from credit_risk.data import load_loans
from credit_risk.split import out_of_time_split
from credit_risk.model import (
    build_logistic,
    build_lgbm,
    build_tree_preprocessor,
    LC_VERDICT_NUMERIC, LC_VERDICT_CATEGORICAL,
    UNDERWRITER_NUMERIC, UNDERWRITER_CATEGORICAL,
)
from credit_risk.evaluate import discrimination_metrics, bootstrap_comparison

TARGET = "target_bad"

df = load_loans()
train, val, test = out_of_time_split(df)

n = len(df)
print(f"train {len(train)/n:.1%}  val {len(val)/n:.1%}  test {len(test)/n:.1%}")

train 53.0%  val 21.8%  test 25.2%


In [2]:
LC_NUM, LC_CAT = LC_VERDICT_NUMERIC, LC_VERDICT_CATEGORICAL
UW_NUM, UW_CAT = UNDERWRITER_NUMERIC, UNDERWRITER_CATEGORICAL

len(UW_NUM + UW_CAT), len(LC_NUM + LC_CAT)

(27, 2)

## Feature sets

The comparison uses three feature sets: borrower information alone, Lending Club's
assessment alone, and their union.

### Underwriter

The underwriter set contains 27 features describing the borrower and requested loan:

- **Loan and income:** `loan_amnt`, `annual_inc`, `dti`, `loan_to_income`.
- **Credit profile:** `fico_range_low`, `credit_history_months`, `inq_last_6mths`,
  `open_acc`, `total_acc`, `active_acct_ratio`, `revol_bal`, `revol_util`.
- **Delinquencies and public records:** `delinq_2yrs`, `mths_since_last_delinq`,
  `delinq_amnt`, `acc_now_delinq`, `pub_rec`, `pub_rec_bankruptcies`,
  `collections_12_mths_ex_med`, `tax_liens`, `chargeoff_within_12_mths`.
- **Application and borrower details:** `home_ownership`, `purpose`, `addr_state`,
  `verification_status`, `application_type`, `emp_length`.

### Lending Club verdict

The verdict set contains two features reflecting Lending Club's assessment:
the assigned interest rate (`int_rate`) and credit grade (`grade`).

### Union

The union combines all 27 underwriter features with `int_rate` and `grade`, giving
29 features. This comparison tests whether borrower information adds predictive
value beyond Lending Club's assessment.

Post-origination variables such as `total_pymnt`, `recoveries` and `last_fico_*`
are excluded to prevent leakage.

## Missing values

Several borrower features contain missing values. For numeric variables, the pipeline replaces
missing values with the median and adds a missing-value indicator. The following comparison
checks whether missingness is associated with a different bad rate for the two features with
the most missing values.

In [3]:
df[UW_NUM + UW_CAT].isnull().sum().loc[lambda s: s > 0].sort_values(ascending=False)

mths_since_last_delinq        359458
emp_length                     42891
pub_rec_bankruptcies             697
revol_util                       404
collections_12_mths_ex_med        56
chargeoff_within_12_mths          56
loan_to_income                    52
tax_liens                         39
dti                               32
active_acct_ratio                  1
dtype: int64

In [4]:
# Summarises missing values and bad rates for one feature.
def missingness(col: str) -> pd.DataFrame:
    status = train[col].isnull().map({True: "missing", False: "present"})
    out = train.groupby(status).agg(loans=(TARGET, "size"), bad_rate=(TARGET, "mean"))
    out["share"] = out["loans"] / out["loans"].sum()
    return out


print(missingness("emp_length"))

print()
print(missingness("mths_since_last_delinq"))

             loans  bad_rate     share
emp_length                            
missing      19816   0.19671  0.052813
present     355396   0.12862  0.947187



                         loans  bad_rate     share
mths_since_last_delinq                            
missing                 199677  0.129134  0.532171
present                 175535  0.135722  0.467829


Rows with missing `emp_length` have a higher bad rate than rows with observed values, so its
missing-value indicator retains information. The bad rates for missing and observed
`mths_since_last_delinq` are similar, suggesting that its indicator contributes less.

## Three feature sets and two models

This section fits logistic regression and LightGBM on each feature set and evaluates them on
validation data:

1. `LC_*`: Lending Club's verdict, used as the benchmark.
2. `UW_*`: borrower data only.
3. Union: borrower data together with Lending Club's verdict.

PR AUC and Brier score are reported alongside ROC AUC. The 15% event rate makes PR AUC part of
the comparison, while the Brier score measures the accuracy of predicted probabilities.
Comparing both models also shows whether nonlinear feature interactions affect performance.

In [5]:
# Fits a model and returns its validation metrics and predictions.
def fit_and_score(
    train: pd.DataFrame,
    val: pd.DataFrame,
    numeric: list[str],
    categorical: list[str],
    build_model: Callable[[list[str], list[str]], Pipeline],
) -> tuple[dict[str, float], NDArray[np.float64]]:
    cols = numeric + categorical
    pipe = build_model(numeric, categorical)
    pipe.fit(train[cols], train[TARGET])

    proba = pipe.predict_proba(val[cols])[:, 1]
    y = val[TARGET].values

    metrics = discrimination_metrics(y, proba)
    return metrics, proba


results_logistic = {
    "lc_verdict":  fit_and_score(train, val, LC_NUM, LC_CAT, build_logistic),
    "underwriter": fit_and_score(train, val, UW_NUM, UW_CAT, build_logistic),
    "union":       fit_and_score(train, val, UW_NUM + LC_NUM, UW_CAT + LC_CAT, build_logistic),
}

results_lgbm = {
    "lc_verdict":  fit_and_score(train, val, LC_NUM, LC_CAT, build_lgbm),
    "underwriter": fit_and_score(train, val, UW_NUM, UW_CAT, build_lgbm),
    "union":       fit_and_score(train, val, UW_NUM + LC_NUM, UW_CAT + LC_CAT, build_lgbm),
}

metrics_logistic = {name: result[0] for name, result in results_logistic.items()}
metrics_lgbm = {name: result[0] for name, result in results_lgbm.items()}

print("logistic")
print(pd.DataFrame(metrics_logistic).T.round(4))

print()
print("lgbm")
print(pd.DataFrame(metrics_lgbm).T.round(4))

logistic
             roc_auc  pr_auc   brier
lc_verdict    0.6787  0.2517  0.1224
underwriter   0.6633  0.2439  0.1224
union         0.6912  0.2665  0.1207

lgbm
             roc_auc  pr_auc   brier
lc_verdict    0.6752  0.2474  0.1217
underwriter   0.6742  0.2538  0.1216
union         0.6967  0.2754  0.1200


The union has the strongest point estimates across the reported metrics. For LightGBM, the
borrower-versus-verdict differences are smaller and vary by metric. The paired comparisons
below assess the precision of these differences and of LightGBM's gain over logistic regression
on the union features.

### Uncertainty in the comparisons

We use a paired bootstrap to estimate uncertainty in the performance differences, since all models score the same validation loans. In each sample, loans are drawn with replacement, keeping each loan's outcome and predictions together across models. The fitted models remain fixed.

Each table shows the difference measured on the full validation set and a 95% confidence interval, given by the 2.5th and 97.5th percentiles of 1,000 bootstrap samples with seed 0. Differences are calculated as first minus second, so positive values favour the first model for ROC AUC and PR AUC, while negative values favour it for Brier score.

The first three tables compare feature sets using LightGBM. The fourth compares LightGBM with logistic regression using the union features.

In [6]:
predictions = {
    "lc_verdict": results_lgbm["lc_verdict"][1],
    "underwriter": results_lgbm["underwriter"][1],
    "union": results_lgbm["union"][1],
    "union_logistic": results_logistic["union"][1],
}
comparisons = {
    "union - lc_verdict": ("union", "lc_verdict"),
    "union - underwriter": ("union", "underwriter"),
    "underwriter - lc_verdict": ("underwriter", "lc_verdict"),
    "union: lgbm - logistic": ("union", "union_logistic"),
}

intervals = bootstrap_comparison(
    val[TARGET], predictions, comparisons, n_bootstrap=1000, seed=0,
)
print(f"{len(val):,} validation loans, 1,000 paired bootstrap samples")
for name in comparisons:
    print(name)
    display(intervals.loc[name].round(4))

154,703 validation loans, 1,000 paired bootstrap samples


union - lc_verdict


,difference,ci_low,ci_high
metric,,,
roc_auc,0.0215,0.0196,0.0238
pr_auc,0.0280,0.0248,0.0314
brier,-0.0017,-0.0019,-0.0015


union - underwriter


,difference,ci_low,ci_high
metric,,,
roc_auc,0.0225,0.0206,0.0244
pr_auc,0.0215,0.0190,0.0240
brier,-0.0016,-0.0017,-0.0014


underwriter - lc_verdict


,difference,ci_low,ci_high
metric,,,
roc_auc,-0.0010,-0.0041,0.0025
pr_auc,0.0064,0.0026,0.0109
brier,-0.0001,-0.0004,0.0001


union: lgbm - logistic


,difference,ci_low,ci_high
metric,,,
roc_auc,0.0054,0.0042,0.0066
pr_auc,0.0089,0.0068,0.0111
brier,-0.0007,-0.0008,-0.0006


The union gives LightGBM a modest but consistent improvement over the verdict alone. ROC AUC rises from 0.6752 to 0.6967, a gain of 0.0215 with a 95% interval of [0.0196, 0.0238]. PR AUC also improves and Brier score decreases, supporting the value of combining the feature sets within this validation period.

LightGBM's ROC AUC gain over logistic regression on the union is smaller: 0.0054 [0.0042, 0.0066], about a quarter of the gain from adding borrower information to the verdict. Here, combining information sources contributes more than changing the model.

Borrower data and the verdict perform similarly when used separately. Their ROC AUC and Brier intervals span zero, while borrower data has a small PR AUC advantage of 0.0064 [0.0026, 0.0109]. Neither feature set clearly dominates across all three metrics.

## Reconstructing the interest rate

To estimate how much of Lending Club's pricing the available borrower features explain, an
`LGBMRegressor` predicts `int_rate` and is evaluated with R^2:

    R^2 = 1 - SS_res / SS_tot     SS_res = sum (y - y_hat)^2,  SS_tot = sum (y - y_bar)^2

R^2 equals 1 for a perfect prediction and 0 when the model performs no better than predicting
the mean interest rate.

In [7]:
reg = Pipeline([
    ("prep", build_tree_preprocessor(UW_NUM, UW_CAT)),
    ("model", LGBMRegressor(n_estimators=300, learning_rate=0.05, num_leaves=31, verbose=-1)),
])
reg.fit(train[UW_NUM + UW_CAT], train["int_rate"])

r2 = r2_score(val["int_rate"], reg.predict(val[UW_NUM + UW_CAT]))
print(f"R^2 reconstructing int_rate: {r2:.2f}")

R^2 reconstructing int_rate: 0.38


The borrower features explain about 40% of the variation in `int_rate`. The remaining variation
may depend on information absent from this dataset, such as credit bureau fields not included
here.

## Risk separation within a grade

The union model's results suggest that borrower variables add information beyond Lending Club's
grade. The following analysis tests whether FICO and DTI separate risk within grade C, which has
enough observations for a within-grade comparison.

In [8]:
grade_c = train[train["grade"] == "C"]

# Summarises bad rates across bands for one feature.
def bad_rate_by_band(col: str, bins: int = 5) -> pd.DataFrame:
    band = pd.qcut(grade_c[col], bins, duplicates="drop")
    return grade_c.groupby(band, observed=True).agg(
        loans=(TARGET, "size"),
        bad_rate=(TARGET, "mean"),
    ).round(3)

print(bad_rate_by_band("fico_range_low"))
print()
print(bad_rate_by_band("dti"))

                  loans  bad_rate
fico_range_low                   
(659.999, 665.0]  22762     0.179
(665.0, 675.0]    23321     0.173
(675.0, 685.0]    18554     0.169
(685.0, 695.0]    12937     0.161
(695.0, 840.0]    17852     0.146

                 loans  bad_rate
dti                             
(-0.001, 10.55]  19091     0.136
(10.55, 15.34]   19087     0.152
(15.34, 19.88]   19113     0.164
(19.88, 25.31]   19055     0.185
(25.31, 39.99]   19080     0.198


Within grade C, bad rates vary across FICO and DTI bands. Higher FICO values and lower DTI
values are associated with lower bad rates, indicating that these variables retain information
not represented by the grade.

## Feature importance

Permutation importance measures how much validation ROC AUC drops when one feature is shuffled.
The model's built-in importance can divide credit between correlated features. Applying
permutation importance to both models on the borrower-only set shows which features each model
uses and whether the engineered ratios and derogatory-event counts contribute.

In [9]:
# Ranks borrower features by permutation importance.
def permutation_ranking(
    build_model: Callable[[list[str], list[str]], Pipeline],
) -> pd.Series:
    cols = UW_NUM + UW_CAT
    model = build_model(UW_NUM, UW_CAT).fit(train[cols], train[TARGET])
    result = permutation_importance(
        model, val[cols], val[TARGET],
        scoring="roc_auc", n_repeats=5, random_state=0, n_jobs=-1,
    )
    return pd.Series(result.importances_mean, index=cols)


importance = pd.DataFrame({
    "logistic": permutation_ranking(build_logistic),
    "lgbm": permutation_ranking(build_lgbm),
})
importance.sort_values("lgbm", ascending=False).head(10).round(4)

,logistic,lgbm
fico_range_low,0.0554,0.0578
revol_bal,0.0037,0.0175
dti,0.0141,0.0150
inq_last_6mths,0.0157,0.0133
loan_to_income,0.0223,0.0093
purpose,0.0113,0.0091
annual_inc,0.0001,0.0067
home_ownership,0.0057,0.0064
loan_amnt,0.0025,0.0053
active_acct_ratio,0.0011,0.0045


In [10]:
FEATURE_CLASSES = {
    "fico": ["fico_range_low"],
    "engineered": ["loan_to_income", "active_acct_ratio"],
    "derogatory": [
        "collections_12_mths_ex_med", "tax_liens", "delinq_amnt", "acc_now_delinq",
        "chargeoff_within_12_mths", "pub_rec", "pub_rec_bankruptcies", "delinq_2yrs",
        "mths_since_last_delinq",
    ],
    "debt_profile": [
        "revol_bal", "revol_util", "dti", "total_acc", "open_acc", "loan_amnt",
        "annual_inc", "credit_history_months",
    ],
    "inquiries": ["inq_last_6mths"],
    "categorical": UW_CAT,
}

pd.DataFrame(
    {name: importance.loc[feats].sum() for name, feats in FEATURE_CLASSES.items()}
).T.round(4)

,logistic,lgbm
fico,0.0554,0.0578
engineered,0.0234,0.0138
derogatory,0.0011,0.0026
debt_profile,0.0235,0.0555
inquiries,0.0157,0.0133
categorical,0.0287,0.0242


FICO has the highest permutation importance for both models, while the remaining features differ
by model.

The engineered ratios contribute more to logistic regression than to LightGBM.
`loan_to_income` supplies an interaction that the linear model cannot form on its own, while the
tree can recover the same signal from the underlying features. LightGBM instead relies more on
the debt-profile variables through nonlinear splits unavailable to the linear model.

The derogatory-event counts contribute little to either model, consistent with their marginal
signal seen earlier. These events are rare and add little once FICO and the debt profile are
included.

## Conclusions

For LightGBM, combining borrower data with the Lending Club verdict gives a modest but consistent improvement across all three metrics, supported by the paired intervals within the validation period. Borrower data and the verdict perform similarly alone, with a small PR AUC advantage for borrower data but uncertain differences in ROC AUC and Brier. LightGBM has higher ROC AUC and PR AUC on the borrower and union sets, while logistic regression scores higher on these metrics for the verdict set, which consists mainly of one monotonic feature.  Adding borrower information to the Lending Club verdict increases LightGBM's ROC AUC by 0.0215. Using LightGBM instead of logistic regression with the same union features gives a smaller gain of 0.0054. In this comparison, combining the information sources matters more than the choice of model.

The borrower variables explain only part of the variation in int_rate, suggesting that some pricing inputs are absent from this dataset. Within grade C, FICO and DTI still separate bad rates, showing that the grade does not capture all the available risk information.

FICO has the highest permutation importance for both models. Beyond FICO, LightGBM relies most on revolving credit balances, DTI and recent credit inquiries, with revolving balances contributing substantially more than in logistic regression. Logistic regression relies more heavily on the engineered loan-to-income ratio. The derogatory-event counts add little to either model once FICO and the other borrower features are included.

The calibration check in scripts/train_baseline.py applies to the baseline union LightGBM model.
